yt_dlp/extractor/yandexvideo.py

```diff
@@ -372,11 +364,17 @@ def _entries(self, item_id, server_state_json, server_settings_json):
     def _real_extract(self, url):
         item_id = self._match_id(url)
         webpage = self._download_webpage(url, item_id)
-        redirect = self._search_json(
-            r'var it\s*=', webpage, 'redirect', item_id, default={}).get('retpath')
-        if redirect:
-            item_id = self._match_id(redirect)
-            webpage = self._download_webpage(redirect, item_id, note='Redirecting')
+        redirect_json = self._search_json(r'var it\s*=', webpage, 'redirect', item_id, default={})
+        retpath = redirect_json.get('retpath')
+        host = redirect_json.get('host')
+        if retpath and host:
+            item_id = self._match_id(retpath)
+            container = self._search_regex(r'element2\.value\s*=\s*\'(.*?)\';', webpage, 'container')
+            _ = self._request_webpage(host, item_id, note='Redirecting', data=urlencode_postdata({
+                'retpath': retpath,
+                'container': container,
+            }))
+            webpage = self._download_webpage(retpath, item_id, note='Downloading video webpage')
```

Comment:
duplicate code. Move into a function

Reply:
Create a base class from which both inherits

Reply:
If creating a base class, we could also generalize the m3u8/mpd format extraction code in `ZenYandexIE` and `YandexVideoIE` and move it into a base class method, something like this:

```py
    def _extract_formats(self, urls, video_id):
        formats = []
        for url in urls:
            if not url_or_none(url):
                continue
            ext = determine_ext(url)
            if ext == 'ismc':
                continue
            elif ext == 'mpd':
                formats.extend(self._extract_mpd_formats(
                    url, video_id, mpd_id='dash', fatal=False))
            elif ext == 'm3u8':
                formats.extend(self._extract_m3u8_formats(
                    url, video_id, 'mp4', m3u8_id='hls', fatal=False))
            else:
                formats.append({'url': url, 'ext': ext})
        return formats
```
The `YandexVideoIE` format extraction code in particular could stand to be simplified somewhat
```diff
-        formats = []
-        streams = content.get('streams') or []
-        streams.append({'url': content.get('content_url')})
-        for stream in streams:
-            content_url = url_or_none(stream.get('url'))
-            if not content_url:
-                continue
-            ext = determine_ext(content_url)
-            if ext == 'ismc':
-                continue
-            elif ext == 'm3u8':
-                formats.extend(self._extract_m3u8_formats(
-                    content_url, video_id, 'mp4',
-                    'm3u8_native', m3u8_id='hls', fatal=False))
-            elif ext == 'mpd':
-                formats.extend(self._extract_mpd_formats(
-                    content_url, video_id, mpd_id='dash', fatal=False))
-            else:
-                formats.append({'url': content_url})
-
-        self._sort_formats(formats)
+        stream_urls = traverse_obj(content, (None, ('content_url', ('streams', ..., 'url'))))
```
then the format extraction could be inlined in the return of both extractors' `_real_extract`:
```py
        return {
            # ...
            'formats': self._extract_formats(stream_urls, video_id),
        }
```
(`self._sort_formats` is no longer needed)

Reply:
Yes. It is expected that code sections will match. YandexVideo and ZenYandex are the same service, the difference is only in the URL part. Need to think.

So far I'm having trouble testing the playlist. Each run fetches a different number of entries and sometimes the test fails. Unclear.